### PR TITLE
client/fuse: set max_idle_threads to the correct value and add support for 3.12 API of libfuse

### PR DESCRIPTION
**Pull request by zhsgao**
I try to use ceph-fuse with libfuse3 and have a problem.
When the version of the libfuse is 3.1 or later, the fuse worker thread of ceph-fuse keeps recreating and deleting file or directory will be blocked forever.
It is caused by parameter 'max_idle_threads' being set to 0 and the relevant logic is in function 'fuse_do_work' of libfuse.
Parameter 'max_idle_threads' can be set by function 'fuse_parse_cmdline', it may be the default value(10 before version 3.12 and -1 after) or a value set by user. It should not be set to 0.

Fixes: https://tracker.ceph.com/issues/58109
Signed-off-by: Zhansong Gao <zhsgao@hotmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by dparmar18** [comment]
@zhsgao Can you please add the tracker link and the explanation to commit message

File: src/client/fuse_ll.cc [hunk 1]
```diff
@@ -1605,9 +1605,10 @@ int CephFuse::Handle::loop()
   if (fuse_multithreaded) {
 #if FUSE_VERSION >= FUSE_MAKE_VERSION(3, 1)
     {
-      struct fuse_loop_config conf = { 0 };
-
-      conf.clone_fd = opts.clone_fd;
+      struct fuse_loop_config conf = {
+        clone_fd: opts.clone_fd,
+        max_idle_threads: opts.max_idle_threads,
```
Comment (dparmar18) [inline src/client/fuse_ll.cc:1610]: nit: there are only two fields in `fuse_loop_config`, maybe the `,`(comma) here can be removed.

**Comment by dparmar18** [comment]
This looks good, did you try test it?

**Comment by zhsgao** [comment]
> @zhsgao Can you please add the tracker link and the explanation to commit message

OK, I will add it later. Thanks!

File: src/client/fuse_ll.cc (hunk 1 above)
Comment (zhsgao) [reply src/client/fuse_ll.cc:1610]: OK, I will remove this comma. Thanks!

**Comment by zhsgao** [comment]
> This looks good, did you try test it?

Yes, I have tested it and it can solve these problems that I found.

**Comment by dparmar18** [comment]
> > This looks good, did you try test it?
> 
> Yes, I have tested it and it can solve these problems that I found.

Awesome. BTW is it possible to share the details of it over here?

**Comment by zhsgao** [comment]
> > > This looks good, did you try test it?
> > 
> > 
> > Yes, I have tested it and it can solve these problems that I found.
> 
> Awesome. BTW is it possible to share the details of it over here?

Yes, I think this problem is very easy to reproduce.
I compile ceph-fuse(v15.2.17) with libfuse(v3.2 or v3.12) in CentOS 8.3 and use it to mount a CephFS.
Then use the command 'top -H -p {pid}' to view all threads of ceph-fuse process.
There are three threads named 'ceph-fuse' and the pid of one of them keeps increasing, so this means a thread keeps recreating.
And then I delete a file or directory and the operation will be blocked forever.
At this time, there are only two threads named 'ceph-fuse' left and there are no fuse worker threads anymore, so the CephFS I mount is unusable.

**Comment by dparmar18** [comment]
> It sounds like fuse is spawning a worker thread for each operation (pid keeps increasing). That sounds expected and would be fixed by your patch. What I find confusing is:
> 
> > At this time, there are only two threads named 'ceph-fuse' left and there are no fuse worker threads anymore, so the CephFS I mount is unusable.
> 
> Why would fuse not spawn a worker thread again for the delete op? I don't see why it became blocked forever?

@zhsgao can you check if `osd blocklist ls` shows any trace? Do you find any backtrace in logs hinting to this?

**Comment by zhsgao** [comment]
> It sounds like fuse is spawning a worker thread for each operation (pid keeps increasing). That sounds expected and would be fixed by your patch. What I find confusing is:
> 
> > At this time, there are only two threads named 'ceph-fuse' left and there are no fuse worker threads anymore, so the CephFS I mount is unusable.
> 
> Why would fuse not spawn a worker thread again for the delete op? I don't see why it became blocked forever?

It may be caused by the logic associated with operation 'FUSE_FORGET' in function 'fuse_do_work' of libfuse. Other operations will decrement the value of variable 'mt->numavail' by 1 and call function 'fuse_loop_start_thread' to start a new worker thread when 'mt->numavail' is 0 before the current thread ends. But this won't happen when the operation is 'FUSE_FORGET' or 'FUSE_BATCH_FORGET'. It looks like this from the code of libfuse and I haven't verified it yet.
I have submitted a PR to libfuse that make it work even if 'max_idle_threads' is set to 0.
https://github.com/libfuse/libfuse/pull/713

**Comment by zhsgao** [comment]
> > It sounds like fuse is spawning a worker thread for each operation (pid keeps increasing). That sounds expected and would be fixed by your patch. What I find confusing is:
> > > At this time, there are only two threads named 'ceph-fuse' left and there are no fuse worker threads anymore, so the CephFS I mount is unusable.
> > 
> > 
> > Why would fuse not spawn a worker thread again for the delete op? I don't see why it became blocked forever?
> 
> @zhsgao can you check if `osd blocklist ls` shows any trace? Do you find any backtrace in logs hinting to this?

I have checked it when blocked and no client is in the blacklist.
I have also check the logs but nothing unusual is found.

**Comment by dparmar18** [comment]
> 

What triggers `FORGET`?

> > It sounds like fuse is spawning a worker thread for each operation (pid keeps increasing). That sounds expected and would be fixed by your patch. What I find confusing is:
> > > At this time, there are only two threads named 'ceph-fuse' left and there are no fuse worker threads anymore, so the CephFS I mount is unusable.
> > 
> > 
> > Why would fuse not spawn a worker thread again for the delete op? I don't see why it became blocked forever?
> 
> It may be caused by the logic associated with operation 'FUSE_FORGET' in function 'fuse_do_work' of libfuse. Other operations will decrement the value of variable 'mt->numavail' by 1 and call function 'fuse_loop_start_thread' to start a new worker thread when 'mt->numavail' is 0 before the current thread ends. But this won't happen when the operation is 'FUSE_FORGET' or 'FUSE_BATCH_FORGET'. It looks like this from the code of libfuse and I haven't verified it yet. I have submitted a PR [libfuse/libfuse#713](https://github.com/libfuse/libfuse/pull/713) to libfuse that make it work even if 'max_idle_threads' is set to 0.

So understanding `FUSE_FORGET` in this scenario where delete operation causes the failure, and If im not wrong then `FORGET` is triggered when kernel shrinks the dentry cache (do mention any other possible situation). So does it mean deleting a file/dir somehows leads to trimming on dcache and thus `FUSE_FORGET`?

**Comment by vshankar** [comment]
This change looks like. @zhsgao According to https://github.com/libfuse/libfuse/pull/713#issuecomment-1332045571, you'd need to change this a bit...

**Comment by zhsgao** [comment]
> > 
> 
> What triggers `FORGET`?
> 
> > > It sounds like fuse is spawning a worker thread for each operation (pid keeps increasing). That sounds expected and would be fixed by your patch. What I find confusing is:
> > > > At this time, there are only two threads named 'ceph-fuse' left and there are no fuse worker threads anymore, so the CephFS I mount is unusable.
> > > 
> > > 
> > > Why would fuse not spawn a worker thread again for the delete op? I don't see why it became blocked forever?
> > 
> > 
> > It may be caused by the logic associated with operation 'FUSE_FORGET' in function 'fuse_do_work' of libfuse. Other operations will decrement the value of variable 'mt->numavail' by 1 and call function 'fuse_loop_start_thread' to start a new worker thread when 'mt->numavail' is 0 before the current thread ends. But this won't happen when the operation is 'FUSE_FORGET' or 'FUSE_BATCH_FORGET'. It looks like this from the code of libfuse and I haven't verified it yet. I have submitted a PR [libfuse/libfuse#713](https://github.com/libfuse/libfuse/pull/713) to libfuse that make it work even if 'max_idle_threads' is set to 0.
> 
> So understanding `FUSE_FORGET` in this scenario where delete operation causes the failure, and If im not wrong then `FORGET` is triggered when kernel shrinks the dentry cache (do mention any other possible situation). So does it mean deleting a file/dir somehows leads to trimming on dcache and thus `FUSE_FORGET`?

I think the failure is caused by FUSE_FORGET and FUSE_FORGET is triggered by delete operation. But I'm not sure how the delete operation triggers FUSE_FORGET. Maybe you are right, it seems like a reasonable explanation.

**Comment by zhsgao** [comment]
> This change looks fine. @zhsgao According to [libfuse/libfuse#713 (comment)](https://github.com/libfuse/libfuse/pull/713#issuecomment-1332045571), you'd need to change this a bit...

The version of the libfuse API currently used ceph-fuse is 3.5 and we can compile it with the latest version(3.12) of libfuse, it is compatible. A new API is provided by libfuse 3.12 and a new option(max_threads) is supported. It can only be set to the default value if we use the old API. We can add support for API 3.12 with a new commit in this PR, or use another PR?

**Comment by vshankar** [comment]
> > This change looks fine. @zhsgao According to [libfuse/libfuse#713 (comment)](https://github.com/libfuse/libfuse/pull/713#issuecomment-1332045571), you'd need to change this a bit...
> 
> The version of the libfuse API currently used ceph-fuse is 3.5 and we can compile it with the latest version(3.12) of libfuse, it is compatible. A new API is provided by libfuse 3.12 and a new option(max_threads) is supported. It can only be set to the default value if we use the old API. We can add support for API 3.12 with a new commit in this PR, or use another PR?

I'll be fine if the changes are included in this PR.

**Comment by zhsgao** [comment]
> > > This change looks fine. @zhsgao According to [libfuse/libfuse#713 (comment)](https://github.com/libfuse/libfuse/pull/713#issuecomment-1332045571), you'd need to change this a bit...
> > 
> > 
> > The version of the libfuse API currently used ceph-fuse is 3.5 and we can compile it with the latest version(3.12) of libfuse, it is compatible. A new API is provided by libfuse 3.12 and a new option(max_threads) is supported. It can only be set to the default value if we use the old API. We can add support for API 3.12 with a new commit in this PR, or use another PR?
> 
> I'll be fine if the changes are included in this PR.

OK, I have done it.

**Review by vshankar — COMMENTED** [review]
LGTM

**Comment by zhsgao** [comment]
> So do we now need the first commit? @zhsgao

Yes, I think so. The first commit is to fix the bug and the second is to support the new API of libfuse. They can be separated.

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#15-Dec-2022

**Pull request merged**